### PR TITLE
Implement Mix-Piece

### LINKED ISSUE #47: Implement Mix-Piece
Implement the Mix-Piece linear approximation algorithm from the paper https://link.springer.com/article/10.1007/s00778-024-00862-z. It is an extension of Sim-Piece proposed in the paper linked in #19 add added to TerseTS by #43.

### PR DESCRIPTION
**Pull request by tomlouisklein**
### Implement Mix-Piece algorithm based on Kitsios, Liakos, Papakonstantinopoulou et.al (2024)

This PR implements the Mix-Piece algorithm from the paper [1] and closes #47. Mix-Piece is a state-of-the-art lossy compression technique for time series data based on Piecewise Linear Approximation (PLA). Mix-Piece significantly outperforms existing PLA algorithms by exploiting similarities among line segments to achieve superior compression ratios while maintaining high accuracy.

The implementation follows the algorithmic description from the paper, with implementatoin details adapted from the existing Sim-Piece algorithm in TerseTS. The authors' reference implementation (https://github.com/xkitsios/Mix-Piece, accessed 08.06.2025) was consulted for validation and clarification of algorithmic details.

**New algorithm**: MixPiece

**Files modified**: 
`src/functional/mix_piece.zig` (implementation of compression and decompression)
`src/tersets.zig` (register new methods)

### Basic functionality (adapted from Sim-Piece)
- Compression/decompression with segment merging
- Even and odd-sized arrays
- Random linear functions with various error bounds

###  Mix-Piece Specific Features

- Time series with trend: Validates improved handling of trending data (key advantage over Sim-Piece)
- Cross-intercept grouping: Tests novel grouping of segments with similar slopes but different starting points
- Dual quantization selection: Verifies floor vs ceiling quantization decision logic
- Edge cases: Single-point segments with very small error bounds

## References:
[1] Kitsios, X., Liakos, P., Papakonstantinopoulou, K., & Kotidis, Y. (2024). Flexible grouping of linear segments for highly accurate lossy compression of time series data. The VLDB Journal, 33, 1569-1589.

### CODE REVIEW

File: src/functional/mix_piece.zig [hunk 1]
```diff
@@ -0,0 +1,1287 @@
...
+const HashMap = std.HashMap;
+
+/// Compresses `uncompressed_values` within `error_bound` using the "Sim-Piece" algorithm.
+/// The function writes the result to `compressed_values`. The `allocator` is used for memory
+/// allocation of intermediate data structures. If an error occurs, it is returned.
+/// This function handles 3 structures subsequently and stores them in 'compressed_values':
+/// Part 1. Segment groups that share the same intercept
+/// Part 2. Segment groups that don't share the same intercept but the same slope
+/// Part 3. Umerged segments
+// Output format: [part1_count][part2_count][part3_count][Part1 data...][Part2 data...][Part3 data...][last_timestamp]
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:48]: We are trying to follow Zig's suggested coding style at https://ziglang.org/documentation/master/#Style-Guide. There you will find that the suggested length of a line is 100 characters. Maybe you can find a way to reduce the size of this line. 
Although overall, this way of documenting this function does not match the rest of the functions. Maybe you can add the information from line 44 ("This function handles ...") as part of the comments of the function instead of the general description of the function.

File: src/functional/mix_piece.zig [hunk 2]
```diff
@@ -0,0 +1,1287 @@
...
+pub fn compressMixPiece(
+    uncompressed_values: []const f64,
+    compressed_values: *ArrayList(u8),
+    allocator: mem.Allocator,
+    error_bound: f32,
+) Error!void {
+    if (error_bound <= 0.0) {
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:55]: Can we think of a way to accept 0.0 as an error bound? This is an open issue that I also have with SimPiece. It would be nice to solve it.

File: src/functional/mix_piece.zig [hunk 3]
```diff
@@ -0,0 +1,1287 @@
...
+        return Error.UnsupportedErrorBound;
+    }
+
+    // Mix-Piece Phase 1: Compute segments metadata
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:59]: Add a . (dot) at the end of any comment.

File: src/functional/mix_piece.zig [hunk 4]
```diff
@@ -0,0 +1,1287 @@
...
+    var segments_metadata = ArrayList(SegmentMetadata).init(allocator);
+    defer segments_metadata.deinit();
+    try computeSegmentsMetadata(
+        uncompressed_values,
+        &segments_metadata,
+        error_bound,
+    );
+
+    // Mix-Piece Phase 2: Merge segments metadata
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:68]: dot at the end.

File: src/functional/mix_piece.zig [hunk 5]
```diff
@@ -0,0 +1,1287 @@
...
+    var cross_intercept_groups = ArrayList(SegmentMetadata).init(allocator);
+    defer cross_intercept_groups.deinit();
+
+    var ungrouped_segments = ArrayList(SegmentMetadata).init(allocator);
+    defer ungrouped_segments.deinit();
+
+    try mergeSegmentsMetadata(segments_metadata, &same_intercept_groups, &cross_intercept_groups, &ungrouped_segments, allocator);
+
+    // Mix-Piece Phase 3: Populate the three data structures
+
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:81]: Why is there a space here? maybe remove!

File: src/functional/mix_piece.zig [hunk 6]
```diff
@@ -0,0 +1,1287 @@
...
+    var cross_intercept_groups = ArrayList(SegmentMetadata).init(allocator);
+    defer cross_intercept_groups.deinit();
+
+    var ungrouped_segments = ArrayList(SegmentMetadata).init(allocator);
+    defer ungrouped_segments.deinit();
+
+    try mergeSegmentsMetadata(segments_metadata, &same_intercept_groups, &cross_intercept_groups, &ungrouped_segments, allocator);
+
+    // Mix-Piece Phase 3: Populate the three data structures
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:80]: dot at the end

File: src/functional/mix_piece.zig [hunk 7]
```diff
@@ -0,0 +1,1287 @@
...
+
+    // Part 1: Same intercept groups
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:82]: dot at the end

File: src/functional/mix_piece.zig [hunk 8]
```diff
@@ -0,0 +1,1287 @@
...
+    var same_intercept_groups = ArrayList(SegmentMetadata).init(allocator);
+    defer same_intercept_groups.deinit();
+
+    var cross_intercept_groups = ArrayList(SegmentMetadata).init(allocator);
+    defer cross_intercept_groups.deinit();
+
+    var ungrouped_segments = ArrayList(SegmentMetadata).init(allocator);
+    defer ungrouped_segments.deinit();
+
+    try mergeSegmentsMetadata(segments_metadata, &same_intercept_groups, &cross_intercept_groups, &ungrouped_segments, allocator);
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:78]: 100 character limit excedeed https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 9]
```diff
@@ -0,0 +1,1287 @@
...
+    var cross_intercept_groups_map = CrossInterceptGroupsMap.init(allocator);
+    defer {
+        var iterator = cross_intercept_groups_map.iterator();
+        while (iterator.next()) |entry| {
+            entry.value_ptr.*.deinit();
+        }
+        cross_intercept_groups_map.deinit();
+    }
+
+    try populateCrossInterceptGroupsAsHashMap(cross_intercept_groups, &cross_intercept_groups_map, allocator);
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:113]: https://ziglang.org/documentation/master/#Style-Guide (from now on I will only comment this issue with the link)

File: src/functional/mix_piece.zig [hunk 10]
```diff
@@ -0,0 +1,1287 @@
...
+    // Store the number of ungrouped segments (Part 3)
+    try appendValue(usize, ungrouped_segments_array.items.len, compressed_values);
+
+    // Now write the actual data for each part
+
+    // Part 1: Same intercept groups
+    try createCompressedRepresentationMergedSegments(same_intercept_groups_map, compressed_values);
+
+    // Part 2: Cross-intercept groups
+    try createCompressedRepresentationCrossInterceptGroups(cross_intercept_groups_map, compressed_values);
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:138]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 11]
```diff
@@ -0,0 +1,1287 @@
...
+
+    // Part 3: Ungrouped segments
+    try createCompressedRepresentationUngroupedSegments(ungrouped_segments_array, compressed_values);
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:141]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 12]
```diff
@@ -0,0 +1,1287 @@
...
+            offset += @sizeOf(f64);
+
+            const slopes_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
+            offset += @sizeOf(f64);
+
+            for (0..slopes_count) |_| {
+                const slope = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+                offset += @sizeOf(f64);
+
+                const timestamps_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:185]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 13]
```diff
@@ -0,0 +1,1287 @@
...
+
+    var offset: usize = 3 * @sizeOf(usize);
+
+    // Part 1. Parse Same Intercept Groups
+    if (part1_count > 0) {
+        // We need to parse part1_count intercept groups
+        // Each group has: intercept, slopes_count, then for each slope: slope, timestamps_count, timestamps
+
+        for (0..part1_count) |_| {
+            const intercept = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:175]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 14]
```diff
@@ -0,0 +1,1287 @@
...
+    const part1_count = header[0]; // Number of intercept groups in Part 1
+    const part2_count = header[1]; // Number of slope groups in Part 2
+    const part3_count = header[2]; // Number of ungrouped segments in Part 3
+
+    var offset: usize = 3 * @sizeOf(usize);
+
+    // Part 1. Parse Same Intercept Groups
+    if (part1_count > 0) {
+        // We need to parse part1_count intercept groups
+        // Each group has: intercept, slopes_count, then for each slope: slope, timestamps_count, timestamps
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:172]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 15]
```diff
@@ -0,0 +1,1287 @@
...
+    // Part 2. Parse Cross Intercept Groups
+    if (part2_count > 0) {
+        // We need to parse part2_count slope groups
+        // Each group has: slope, pair_count, then for each pair: intercept, timestamp_delta
+
+        for (0..part2_count) |_| {
+            const slope = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+            offset += @sizeOf(f64);
+
+            const pair_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:214]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 16]
```diff
@@ -0,0 +1,1287 @@
...
+            for (0..slopes_count) |_| {
+                const slope = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+                offset += @sizeOf(f64);
+
+                const timestamps_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
+                offset += @sizeOf(f64);
+
+                var timestamp: usize = 0;
+                for (0..timestamps_count) |_| {
+                    const delta = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:190]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 17]
```diff
@@ -0,0 +1,1287 @@
...
+
+        for (0..part1_count) |_| {
+            const intercept = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+            offset += @sizeOf(f64);
+
+            const slopes_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
+            offset += @sizeOf(f64);
+
+            for (0..slopes_count) |_| {
+                const slope = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:182]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 18]
```diff
@@ -0,0 +1,1287 @@
...
+    // Part 1. Parse Same Intercept Groups
+    if (part1_count > 0) {
+        // We need to parse part1_count intercept groups
+        // Each group has: intercept, slopes_count, then for each slope: slope, timestamps_count, timestamps
+
+        for (0..part1_count) |_| {
+            const intercept = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+            offset += @sizeOf(f64);
+
+            const slopes_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:178]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 19]
```diff
@@ -0,0 +1,1287 @@
...
+
+            const pair_count = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
+            offset += @sizeOf(f64);
+
+            var timestamp: usize = 0;
+            for (0..pair_count) |_| {
+                const intercept = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+                offset += @sizeOf(f64);
+
+                const delta = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:222]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 20]
```diff
@@ -0,0 +1,1287 @@
...
+        // Each segment has: slope, intercept, timestamp_delta
+        var timestamp: usize = 0;
+        for (0..part3_count) |_| {
+            const slope = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+            offset += @sizeOf(f64);
+
+            const intercept = mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0];
+            offset += @sizeOf(f64);
+
+            const delta = @as(usize, @bitCast(mem.bytesAsSlice(f64, compressed_values[offset .. offset + @sizeOf(f64)])[0]));
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:248]: https://ziglang.org/documentation/master/#Style-Guide

File: src/functional/mix_piece.zig [hunk 21]
```diff
@@ -0,0 +1,1287 @@
+// Copyright 2025 TerseTS Contributors
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:1]: Overall, if the number of shared functions is significant, maybe we should consider putting Sim-Piece and Mix-Piece in the same file or simply importing those functions from Sim-Piece instead of rewriting them again here. Maybe consider `sp = @import('src/sim_piece.zig')` at the beginning and reusing as much as possible.

File: src/functional/mix_piece.zig [hunk 22]
```diff
@@ -0,0 +1,1287 @@
...
+    try populateCrossInterceptGroupsAsHashMap(cross_intercept_groups, &cross_intercept_groups_map, allocator);
+
+    // Part 3: Ungrouped segments
+    var ungrouped_segments_array = ArrayList(UngroupedSegment).init(allocator);
+    defer ungrouped_segments_array.deinit();
+
+    try populateUngroupedSegmentsArray(ungrouped_segments, &ungrouped_segments_array);
+
+    // Mix-Piece Phase 4: Create compressed representation with boundaries
+
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:122]: Why is there space here? Consider removing it!

File: src/functional/mix_piece.zig [hunk 23]
```diff
@@ -0,0 +1,1287 @@
...
+    try appendValue(usize, same_intercept_groups_map.count(), compressed_values);
+
+    // Store the number of slope groups (Part 2)
+    try appendValue(usize, cross_intercept_groups_map.count(), compressed_values);
+
+    // Store the number of ungrouped segments (Part 3)
+    try appendValue(usize, ungrouped_segments_array.items.len, compressed_values);
+
+    // Now write the actual data for each part
+
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:133]: Why is there space here? Consider removing it!

File: src/functional/mix_piece.zig [hunk 24]
```diff
@@ -0,0 +1,1287 @@
...
+    try createCompressedRepresentationCrossInterceptGroups(cross_intercept_groups_map, compressed_values);
+
+    // Part 3: Ungrouped segments
+    try createCompressedRepresentationUngroupedSegments(ungrouped_segments_array, compressed_values);
+
+    // The last timestamp must be stored, otherwise the end time during decompression is unknown.
+    try appendValue(usize, uncompressed_values.len, compressed_values);
+}
+
+/// Decompress `compressed_values` produced by "Mix-Piece".
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:147]: Maybe simply replicate the style used when comment the decompress function of sim-piece.

File: src/functional/mix_piece.zig [hunk 25]
```diff
@@ -0,0 +1,1287 @@
...
+    compressed_values: []const u8,
+    decompressed_values: *ArrayList(f64),
+    allocator: mem.Allocator,
+) Error!void {
+
+    // Initialize temp array to store all segments
+    var all_segments = ArrayList(SegmentMetadata).init(allocator);
+    defer all_segments.deinit();
+
+    // Read header to get structure counts (not sizes!)
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:161]: Not clear what (not sizes!) means?

File: src/functional/mix_piece.zig [hunk 26]
```diff
@@ -0,0 +1,1287 @@
...
+/// Uses both floor and ceil quantization to find optimal segments.
+fn computeSegmentsMetadata(
+    uncompressed_values: []const f64,
+    segments_metadata: *ArrayList(SegmentMetadata),
+    error_bound: f32,
+) Error!void {
+    // Adjust the error bound to avoid exceeding it during decompression.
+    const adjusted_error_bound = error_bound - shared.ErrorBoundMargin;
+
+    // Track bounds for floor quantization
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:300]: Please finish every comment with a . at the end! Check everywhere.

File: src/functional/mix_piece.zig [hunk 27]
```diff
@@ -0,0 +1,1287 @@
...
+/// This algorithm has two parts:
+/// 1. Group segments with the same starting point
+/// 2. Group remaining ungrouped segments regardless of starting point
+///
+/// The results are stored in three output arrays:
+/// - `same_intercept_groups`: Groups formed from segments sharing the same quantized intercept value
+/// - `cross_intercept_groups`: Groups formed from segments with different intercept values
+/// - `ungrouped_segments`: Segments that couldn't be grouped with any other segment
+/// The `allocator` is used to allocate memory for the intermediate representations needed.
+fn mergeSegmentsMetadata(
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:501]: Is this mergeSegment the same as Sim-Piece? If so, maybe just reuse?

File: src/functional/mix_piece.zig [hunk 28]
```diff
@@ -0,0 +1,1287 @@
...
+                    .upper_bound_slope = upper_bound_slope_floor,
+                    .lower_bound_slope = lower_bound_slope_floor,
+                });
+            }
+        }
+    }
+}
+
+/// Mix-Piece Phase 2. Merge the elements in `segments_metadata` using Algorithm 4 from the paper.
+/// This algorithm has two parts:
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:492]: Normally, in the overall comment of a function, there is only information about the signature of the function and the output. Comments about the logic are embedded inside the function to explain the lines of code. Maybe you can fix that everywhere in the general comment of the functions implemented. Moreover, if you see that we violate that principle anywhere else, feel free to highlight it or create an issue so we can take care of it.

File: src/functional/mix_piece.zig [hunk 29]
```diff
@@ -0,0 +1,1287 @@
...
+    for (segments_metadata.items) |segment_metadata| {
+        try appendSegmentMetadata(
+            &segments_by_intercept,
+            segment_metadata,
+            allocator,
+        );
+    }
+
+    // 1. Group segments with the same starting point
+
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:537]: Maybe remove the space?

File: src/functional/mix_piece.zig [hunk 30]
```diff
@@ -0,0 +1,1287 @@
...
+                .intercept = current_group.intercept,
+                .lower_bound_slope = current_group.lower_bound_slope,
+                .upper_bound_slope = current_group.upper_bound_slope,
+            });
+        }
+        timestamps_array.clearRetainingCapacity();
+    }
+
+    // 2. Group remaining ungrouped segments across different intercept values
+
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:633]: Maybe remove the empty line

File: src/functional/mix_piece.zig [hunk 31]
```diff
@@ -0,0 +1,1287 @@
...
+/// Mix-Piece Phase 3.2. Populate HashMap of cross_intercept_groups
+/// Groups segments by their slope and stores intercept-timestamp pairs
+/// The `allocator` is used to allocate memory of intermediates.
+/// Output format: slope -> ArrayList(InterceptTimestampPair)
+/// Which corresponds to: a_j; k_j; b_j,1; t_j,1; b_j,k_j; t_j,k_j
+/// where:
+/// - a_j = slope
+/// - k_j = number of intercept-timestamp pairs
+/// - b_j,i; t_j,i = intercept and timestamp pairs
+fn populateCrossInterceptGroupsAsHashMap(
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:806]: Is this function the same as in Sim-Piece's `populateSegmentsMetadataHashMap`? If so, consider reuse. Moreover, try to reproduce the style of the comment of that function.

File: src/functional/mix_piece.zig [hunk 32]
```diff
@@ -0,0 +1,1287 @@
...
+                previous_timestamp = timestamp;
+            }
+        }
+    }
+}
+
+/// Mix-Piece Phase 4.2. Create compressed representation from the 'cross_intercept_groups_map'
+/// Output format: For each slope a_j, store the count k_j of intercept-timestamp pairs,
+/// followed by the pairs themselves: [a_j, k_j, b_j,1, t_j,1, b_j,2, t_j,2, ..., b_j,k_j, t_j,k_j]
+/// where:
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:905]: This style of using: `where: `, does not really match the style used across TerseTS. Maybe you can modify everywhere to maintain a concise comment style across the whole library.

File: src/functional/mix_piece.zig [hunk 33]
```diff
@@ -0,0 +1,1287 @@
...
+    // Use the average of upper and lower slopes (in our case they're the same)
+    const slope = (segment_metadata.upper_bound_slope + segment_metadata.lower_bound_slope) / 2.0;
+
+    for (start_time..end_time) |t| {
+        const decompressed_value = slope * @as(f64, @floatFromInt(t - start_time)) + segment_metadata.intercept;
+        try decompressed_values.append(decompressed_value);
+    }
+}
+
+// Tests
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:1101]: Maybe remove, is not necessary.

File: src/functional/mix_piece.zig [hunk 34]
```diff
@@ -0,0 +1,1287 @@
...
+test "mix-piece can compress, decompress and merge many segments with non-zero error bound" {
+    const seed: u64 = @bitCast(time.milliTimestamp());
+    var prng = rand.DefaultPrng.init(seed);
+    const random = prng.random();
+    const error_bound = 0.01;
+    const allocator = testing.allocator;
+    var uncompressed_values = ArrayList(f64).init(allocator);
+    defer uncompressed_values.deinit();
+    for (0..20) |_| {
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:1110]: Maybe add a line between this two lines to make look better.

File: src/functional/mix_piece.zig [hunk 35]
```diff
@@ -0,0 +1,1287 @@
...
+        try decompressed_values.append(decompressed_value);
+    }
+}
+
+// Tests
+test "mix-piece can compress, decompress and merge many segments with non-zero error bound" {
+    const seed: u64 = @bitCast(time.milliTimestamp());
+    var prng = rand.DefaultPrng.init(seed);
+    const random = prng.random();
+    const error_bound = 0.01;
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:1106]: Maybe make the error bound also random between 0 and 1. One of the opened PRs is already changing that in Sim-Piece as well. Take a look at #79, the `sim_piece.zig` file.

File: src/functional/mix_piece.zig [hunk 36]
```diff
@@ -0,0 +1,1287 @@
...
+    const allocator = testing.allocator;
+    var uncompressed_values = ArrayList(f64).init(allocator);
+    defer uncompressed_values.deinit();
+
+    // Generate data with a clear upward trend
+    for (0..100) |i| {
```
Comment (cmcuza) [inline src/functional/mix_piece.zig:1194]: Take a look at #79 to replicate the why to generate random numbers. Although I am not sure if they can be used already. Take a look!

File: src/functional/mix_piece.zig (hunk 18 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:213]: done

File: src/functional/mix_piece.zig (hunk 17 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:217]: done

File: src/functional/mix_piece.zig (hunk 12 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:220]: done

File: src/functional/mix_piece.zig (hunk 16 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:225]: done

File: src/functional/mix_piece.zig (hunk 15 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:249]: done

File: src/functional/mix_piece.zig (hunk 19 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:257]: done

File: src/functional/mix_piece.zig (hunk 20 above)
Comment (tomlouisklein) [reply src/functional/mix_piece.zig:283]: done

**Review by cmcuza — APPROVED** [review]
After editing myself a few minor things, the PR is ready to merge.

**Review by cmcuza — APPROVED** [review]
